src/mouse_events.cpp

```diff
@@ -785,17 +789,26 @@ void mouse_handler::select_or_action(bool browse)
 	wb::future_map_if_active planned_unit_map;
 
 	if(game_lua_kernel* lk = pc_.gamestate().lua_kernel_.get()) {
-		lk->select_hex_callback(last_hex_);
+		lk->select_hex_callback(clicked_hex);
+		// The callback may modify clicked_hex, so make sure it's still valid
+		if(!pc_.get_map().on_board(clicked_hex)) {
+			return;
+		}
 	}
 
-	unit_map::iterator clicked_u = find_unit(last_hex_);
+	unit_map::iterator clicked_u = find_unit(clicked_hex);
 	unit_map::iterator selected_u = find_unit(selected_hex_);
 
 	if(clicked_u && (!selected_u || selected_u->side() != side_num_ ||
 	  (clicked_u->side() == side_num_ && clicked_u->id() != selected_u->id()))
 	) {
-		select_hex(last_hex_, false);
+		select_hex(clicked_hex, false);
 	} else {
+		// There's a small chance the last hex has changed since this function began.
+		// Make sure the move applies to the hex that was originally clicked!
+		utils::guard_value clicked(last_hex_, clicked_hex);
+		utils::guard_value prev(previous_hex_, attack_hex);
+		utils::guard_value route(current_route_, saved_route);
```

Comment:
Looks like `move_action` is so complex that the easiest way to handle this is to still call it, but override three of its side effects after the call. My instinct for this part of the codebase is that that's probably true, but it's also why I'm so adverse to additional complexity in this area.

It's going to take me a lot of time to review these three lines' interaction with the other side effects of `move_action`, I can't commit to it this week.

Reply:
@Pentarctagon, thanks for the offer to help with the reviews. I feel that the way the unit selection and `previous_hex`, `clicked_hex`, `attack_hex`, `saved_route`, etc are used in the existing code is spaghetti, and that adding `utils::guard_value` around the call to `move_action` is making the spaghetti worse. Please could I offload to you the task of working out if this is a good change, and maybe working out if there's a long-term plan to improve this?

Reply:
Would you be able to explain more what you mean by "Looks like move_action is so complex that the easiest way to handle this is to still call it, but override three of its side effects after the call"?

This doesn't seem like a result of complexity, but simply that no matter how complex it is or how long it takes to execute, there's a chance someone manages to click something else half way through.

Reply:
Sorry for taking so long to respond to this:

I'm not complaining about the "make sure the move applies to the hex that was originally clicked", I'm complaining that:
* The three values overwritten before calling `move_action` are a subset of the ones that control `move_action`.
* After `move_action` returns, a subset of the values that it changes are overwritten by the destructors of `utils::guard_value`.
* My gut feeling is that the new code isn't covering all edge-cases of the existing code.

@CelticMinstrel, why is `guard_value` used here? If `move_action` changes the values, why should this overwrite them after it returns? Why are these three specific member variables overwritten, but no others?

Walking through `move_action`:
* There are a lot of edge-cases, testing is unlikely to even cover all of the ones that can happen without the whiteboard.
* There's logic using `selected_hex_`, which may have changed since the start of `select_or_action` because `mouse_motion` can change it. In the current PR's code there's some checks about which team the unit (if any) on `selected_hex_` is on, but that still leaves a wide possibility for `selected_hex_` to have changed since the start of `select_or_action`.
* There's logic involving `last_hex_` and `selected_hex_`, where exactly one of those two has been overwritten in `select_or_action` to have the value that it had when `select_or_action` was called.
* Several paths through `move_action` call `select_hex`, a function which can change many of the class' member variables, a superset of the ones that will be restored by the `guard_value` destructors.

Reply:
> Why are these three specific member variables overwritten, but no others?

They are the ones I identified as being necessary to overwrite. If I need to add another one, of course I can do that.

Reply:
Please share your knowledge about why that's the necessary set - that knowledge would be really helpful for reviewing.

Reply:
From what I could tell, these are the variables that determine what action is taken as a result of the click. Under normal circumstances, they'd never change during the execution of this function, but if the callback takes some time to execute, they _might_ change… which was what this was supposed to fix.

Reply:
My analysis on May 12th shows that your set is incomplete, there are more variables that determine what action is taken as a result of the click. However, that doesn't mean my analysis is the full set, it's just sufficient to review and say that you, as PR author, need to do a complete analysis.

You've also only responded to a single part of my questions from May 12th. If `move_action` changes the values, why should this overwrite them after it returns?

Reply:
Isn't the reason for that the whole reason this PR was opened in the first place? To fix the issue of the action triggering on the wrong hex if the player moved the mouse?